### PR TITLE
Should add a support of headers in a client and a server side

### PR DESCRIPTION
**Pull request by ozonophore**

### CODE REVIEW

File: generators/java/reactor-netty/expected-client/src/test/java/com/example/routes/SimpleClientTest.java [hunk 1]
```diff
@@ -44,7 +46,7 @@ public void testPost() {
     @Test
     @Override
     public void testGet() {
-        Mono<Dto> response = simpleClient.simpleGet("idValue", "param1Value", Param2OfSimpleGet.VALUE1);
+        Mono<Dto> response = simpleClient.simpleGet(LocalDate.now(), "idValue", "param1Value", Param2OfSimpleGet.VALUE1);
```
Comment (andrey-fomin) [inline generators/java/reactor-netty/expected-client/src/test/java/com/example/routes/SimpleClientTest.java:49]: minor. It's better to constant values in tests (not random or current time-based)

File: generators/java/reactor-netty/generator/src/main/kotlin/io/github/fomin/oasgen/java/reactor/netty/ReactorNettyClientWriter.kt [hunk 2]
```diff
@@ -93,10 +93,29 @@ class ReactorNettyClientWriter(
...
+                            if (parameter.schemaParameter.parameterIn == ParameterIn.HEADER) {
+                                val stringWriteExpression =
+                                        converterRegistry[parameter.schemaParameter.schema()].stringWriteExpression("param$index")
+                                """.add("${parameter.name}", param$index != null ? $stringWriteExpression : null)"""
+                            } else {
+                                null
+                            }
+                        }
+                val headerArgsBlock = when(headerArgs.isNotEmpty()) {
+                    true -> """|        .headers(headers -> headers${headerArgs.indentWithMargin(5)})""".trimMargin()
```
Comment (andrey-fomin) [inline generators/java/reactor-netty/generator/src/main/kotlin/io/github/fomin/oasgen/java/reactor/netty/ReactorNettyClientWriter.kt]: strange formatting...
margin does not make sense for single line literals.

why `indentWithMargin(5)` is used here?

File: generators/java/reactor-netty/generator/src/main/kotlin/io/github/fomin/oasgen/java/reactor/netty/ReactorNettyClientWriter.kt [hunk 3]
```diff
@@ -93,10 +93,29 @@ class ReactorNettyClientWriter(
                     }
                 }.joinToString("")
 
-                val parameterStrDeclarations = javaOperation.parameters.mapIndexed { index, javaParameter ->
-                    val stringWriteExpression =
-                            converterRegistry[javaParameter.schemaParameter.schema()].stringWriteExpression("param$index")
-                    "String param${index}Str = param$index != null ? $stringWriteExpression : null;"
+                val parameterStrDeclarations = javaOperation.parameters.mapIndexedNotNull { index, javaParameter ->
+                            if (javaParameter.schemaParameter.parameterIn != ParameterIn.HEADER) {
```
Comment (andrey-fomin) [inline generators/java/reactor-netty/generator/src/main/kotlin/io/github/fomin/oasgen/java/reactor/netty/ReactorNettyClientWriter.kt:97]: Why headers are excluded here? What for they are handled specially in next block?

File: generators/java/spring-web/expected-client/src/test/java/com/example/SimpleClientTest.java [hunk 4]
```diff
@@ -74,7 +76,7 @@ public void testPost() {
     @Override
     @Test
     public void testGet() {
-        Dto response = simpleClient.simpleGet("idValue", "param1Value", Param2OfSimpleGet.VALUE1);
+        Dto response = simpleClient.simpleGet(LocalDate.now(), "idValue", "param1Value", Param2OfSimpleGet.VALUE1);
```
Comment (andrey-fomin) [inline generators/java/spring-web/expected-client/src/test/java/com/example/SimpleClientTest.java:79]: Please don't use current time-based values in tests

File: generators/java/spring-web/generator/src/main/kotlin/io/github/fomin/oasgen/java/spring/web/JavaSpringMvcServerWriter.kt [hunk 5]
```diff
@@ -112,26 +113,45 @@ class JavaSpringMvcServerWriter(
...
-                        ParameterIn.QUERY -> """request.getParameter("${parameter.name}")"""
-                    }
-                    val converterWriter = converterRegistry[schema]
-                    """${converterWriter.valueType()} param$index = ${
-                        converterWriter.stringParseExpression(
-                            parameterExpression
-                        )
-                    };"""
+                val parameterDefinitions = operation.parameters()
+                        .mapIndexedNotNull { index, parameter ->
```
Comment (andrey-fomin) [inline generators/java/spring-web/generator/src/main/kotlin/io/github/fomin/oasgen/java/spring/web/JavaSpringMvcServerWriter.kt]: mapIndexedNotNull - why not null?

File: generators/java/spring-web/generator/src/main/kotlin/io/github/fomin/oasgen/java/spring/web/JavaSpringMvcServerWriter.kt [hunk 6]
```diff
@@ -112,26 +113,45 @@ class JavaSpringMvcServerWriter(
                 } else {
                     ""
                 }
+                val headerDefinitions = operation.parameters()
```
Comment (andrey-fomin) [inline generators/java/spring-web/generator/src/main/kotlin/io/github/fomin/oasgen/java/spring/web/JavaSpringMvcServerWriter.kt]: I don't see the reason to define header-parameters separately. They can be defined with the other parameters

**Review by andrey-fomin — COMMENTED** [review]
I don't see changes in class ReferenceServer.
It should check incoming header value

File: generators/java/reactor-netty/generator/src/main/kotlin/io/github/fomin/oasgen/java/reactor/netty/ReactorNettyClientWriter.kt (hunk 3 above)
Comment (ozonophore) [reply generators/java/reactor-netty/generator/src/main/kotlin/io/github/fomin/oasgen/java/reactor/netty/ReactorNettyClientWriter.kt:97]: > Why headers are excluded here? What for they are handled specially in next block?

Anyway, it should be a new block to add parameters to the header's list

File: generators/java/reactor-netty/generator/src/main/kotlin/io/github/fomin/oasgen/java/reactor/netty/ReactorNettyClientWriter.kt (hunk 2 above)
Comment (ozonophore) [reply generators/java/reactor-netty/generator/src/main/kotlin/io/github/fomin/oasgen/java/reactor/netty/ReactorNettyClientWriter.kt]: > strange formatting...
> margin does not make sense for single line literals.
> 
> why `indentWithMargin(5)` is used here?

I am not sure... but I can't do that several parameters of headers have the same gaps

File: generators/java/spring-web/generator/src/main/kotlin/io/github/fomin/oasgen/java/spring/web/JavaSpringMvcServerWriter.kt (hunk 6 above)
Comment (ozonophore) [reply generators/java/spring-web/generator/src/main/kotlin/io/github/fomin/oasgen/java/spring/web/JavaSpringMvcServerWriter.kt]: > I don't see the reason to define header-parameters separately. They can be defined with the other parameters

changed it

File: generators/java/spring-web/generator/src/main/kotlin/io/github/fomin/oasgen/java/spring/web/JavaSpringMvcServerWriter.kt (hunk 5 above)
Comment (ozonophore) [reply generators/java/spring-web/generator/src/main/kotlin/io/github/fomin/oasgen/java/spring/web/JavaSpringMvcServerWriter.kt]: > mapIndexedNotNull - why not null?

returned it back

File: generators/java/reactor-netty/generator/src/main/kotlin/io/github/fomin/oasgen/java/reactor/netty/ReactorNettyClientWriter.kt (hunk 2 above)
Comment (ozonophore) [reply generators/java/reactor-netty/generator/src/main/kotlin/io/github/fomin/oasgen/java/reactor/netty/ReactorNettyClientWriter.kt]: > strange formatting...
> margin does not make sense for single line literals.
> 
> why `indentWithMargin(5)` is used here?

Added a little change of format. indentWithMargin(5) was added to move internal block to right

**Comment by andrey-fomin** [comment]
fixed in cfbd7d342c629ddbfcd6ec70b212d2d86af38a7d

**Pull request closed without merging**